### PR TITLE
fix(import routes): merge route when route have the same name

### LINKED ISSUE #2329: 导入OpenAPI会切分多个URI路径
### Issue description

同一个路由若有多个路径信息，导出后再导入会生成多个路由

### Expected behavior

只生成一个路由

### How to Reproduce

导出前
导出截图

导入后
导入截图

### Screenshots

_No response_

### Environment

- apisix version (cmd: `apisix version`):
- OS (cmd: `uname -a`):
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
- etcd version, if have (cmd: run `etcd --version`):
- apisix-dashboard version, if have:
- Browser version, if have:

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by kevinw66**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?** 

- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

Please update this section with detailed description.

**Related issues**

fixes #2329

**Checklist:**

- [ ] Did you explain what problem does this PR solve? Or what new features have been added?
- [ ] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [ ] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

**Comment by bzp2010** [comment]
Hi, I'm not sure if it's appropriate to merge based on `name`, as far as I understand it, ID should be the basis for merging, and sometimes even when there is overlapping content, no data should be merged.

**Comment by kevinw66** [comment]
> 

When creating a new route, the name cannot be the same, or else it will throw an error.
So I merge routes based on name, Because duplicated name was treated as dirty data.

**Comment by zaunist** [comment]
> > Hi, I'm not sure if it's appropriate to merge based on `name`, as far as I understand it, ID should be the basis for merging, and sometimes even when there is overlapping content, no data should be merged.
> 
> @bzp2010 When creating a new route, the name cannot be the same, or else it will throw an error. So I merge routes based on name, Because duplicated name was treated as dirty data.

Could we use based on ID ? I think ID is better.

**Comment by Baoyuantop** [comment]
Agree with ID

**Comment by kevinw66** [comment]
If we merge based on id, the bug described in issue cannot be fixed, the route with same name and different path value already assigned different id

**Comment by jwrookie** [comment]
> If we merge based on id, the bug described in issue cannot be fixed, the route with same name and different path value already assigned different id unless we cache id with route name, and if we meet same route name, we use the same id. but what's the different between this and merge directly with name ?

I think both import and export should be done through `ID`

**Comment by kevinw66** [comment]
@bzp2010 @zaunist @Baoyuantop What do you think?

**Comment by Baoyuantop** [comment]
Hi @kevinw66, maybe we can discuss this with more people in the community mail list, here is the subscription method: https://apisix.apache.org/docs/general/community#get-help-from-mailing-list-recommended

**Comment by zaunist** [comment]
> Hi @kevinw66, maybe we can discuss this with more people in the community mail list, here is the subscription method: https://apisix.apache.org/docs/general/community#get-help-from-mailing-list-recommended

Agree. We need to discuss with others.

**Review by nic-chen — COMMENTED** [review]
@kevinw66 hi
need some test cases for the change

File: api/internal/handler/data_loader/route_import.go [hunk 1]
```diff
@@ -137,6 +137,22 @@ func (h *ImportHandler) Import(c droplet.Context) (interface{}, error) {
...
 	}
 
+	// merge route
+	idRoute := make(map[string]*entity.Route)
+	for _, route := range routes {
+		if existRoute, ok := idRoute[route.RouteId]; ok {
+			uris := append(existRoute.Uris, route.Uris...)
+			existRoute.Uris = uris
+		} else {
+			idRoute[route.RouteId] = route
```
Comment (nic-chen) [inline api/internal/handler/data_loader/route_import.go:147]: It seems that the combination of `method` is not considered.

Comment (kevinw66) [reply api/internal/handler/data_loader/route_import.go:147]: `methods` may already merged by `OpenAPI3ToRoute` method via path, and route have different path become different routes.
Here I just merge path via original route id.

File: api/test/e2e/route_import_test.go [hunk 2]
```diff
@@ -577,3 +577,136 @@ func TestRoute_export_import(t *testing.T) {
...
+			Desc:   "Create a route",
+			Object: ManagerApiExpect(t),
+			Method: http.MethodPut,
+			Path:   "/apisix/admin/routes/r1",
+			Body: `{
+					"id": "r1",
+					"uris": ["/test1", "/test2"],
+					"name": "route_all",
+					"desc": "所有",
+					"methods": ["GET"],
```
Comment (nic-chen) [inline api/test/e2e/route_import_test.go:594]: Multiple methods should be set up to verify that the route is split after exporting, and verify the exported data is merged into one route after reimporting.

Comment (kevinw66) [reply api/test/e2e/route_import_test.go]: Done

File: api/internal/core/entity/entity.go [hunk 3]
```diff
@@ -85,6 +85,7 @@ type Route struct {
 	ServiceProtocol string                 `json:"service_protocol,omitempty"`
 	Labels          map[string]string      `json:"labels,omitempty"`
 	EnableWebsocket bool                   `json:"enable_websocket,omitempty"`
+	RouteId         string                 `json:"route_id,omitempty"`
```
Comment (nic-chen) [inline api/internal/core/entity/entity.go:88]: we don't need this field

File: api/internal/handler/data_loader/route_import.go [hunk 4]
```diff
@@ -137,6 +137,26 @@ func (h *ImportHandler) Import(c droplet.Context) (interface{}, error) {
...
+	for _, route := range routes {
+		if route.RouteId == "" {
+			route.RouteId = route.ID.(string)
+		}
+		//
+		if existRoute, ok := idRoute[route.RouteId]; ok {
+			uris := append(existRoute.Uris, route.Uris...)
+			existRoute.Uris = uris
+		} else {
+			idRoute[route.RouteId] = route
```
Comment (nic-chen) [inline api/internal/handler/data_loader/route_import.go:151]:
```suggestion
			idRoute[route.ID] = route
```

**Comment by nic-chen** [comment]
FE test failed. @Baoyuantop @guoqqqi @bzp2010 please help when you have time, thanks.

**Comment by guoqqqi** [comment]
image
Hi, @kevinw66  looks like a new id field has been added? We need to update the `web/cypress/fixtures/export-route-dataset.json` file

**Comment by kevinw66** [comment]
> image Hi, @kevinw66 looks like a new id field has been remove? We need to update the `web/cypress/fixtures/export-route-dataset.json` file

Is this id a fixed value?

**Comment by guoqqqi** [comment]
As the ID is not a fixed value, we need to do some processing. We need to remove the id field from the `fileContent`.
https://github.com/apache/apisix-dashboard/blob/master/web/cypress/integration/route/import_export_route.spec.js#L134
If you need help please leave me a message 😄

**Comment by kevinw66** [comment]
> As the ID is not a fixed value, we need to do some processing. We need to remove the id field from the `fileContent`. https://github.com/apache/apisix-dashboard/blob/master/web/cypress/integration/route/import_export_route.spec.js#L134 If you need help please leave me a message 😄

Of course, It's hard for me to change frontend code 😢

**Comment by guoqqqi** [comment]
cc @oil-oil can you help to fix the E2E error?

**Review by nic-chen — APPROVED** [review]

**Review by bzp2010 — APPROVED** [review]

**Pull request merged**